llama.cpp

```diff
@@ -985,77 +970,131 @@ static void llama_model_load_internal(
+                    + ggml_nbytes(layer.w1) + ggml_nbytes(layer.w2) + ggml_nbytes(layer.w3);
+            }
         }
     }
 
     ml->done_getting_tensors();
 
+    // print memory requirements
+    {
+        const size_t scale = memory_type == GGML_TYPE_F32 ? 2 : 1;
+
+        // this is the total memory required to run the inference
+        const size_t mem_required =
+            ctx_size +
+            mmapped_size - vram_total + // weights in VRAM not in memory
+            MEM_REQ_SCRATCH0().at(model.type) +
+            MEM_REQ_SCRATCH1().at(model.type) +
+            MEM_REQ_EVAL().at(model.type);
+
+        // this is the memory required by one llama_state
+        const size_t mem_required_state =
+            scale*MEM_REQ_KV_SELF().at(model.type);
+
+        fprintf(stderr, "%s: mem required  = %7.2f MB (+ %7.2f MB per state)\n", __func__,
+                mem_required / 1024.0 / 1024.0, mem_required_state / 1024.0 / 1024.0);
+
+#ifdef GGML_USE_CUBLAS
+        const int n_gpu = std::min(n_gpu_layers, int(hparams.n_layer));
+
+        fprintf(stderr, "%s: [cublas] offloading %d layers to GPU\n", __func__, n_gpu);
+        if (n_gpu_layers > (int) hparams.n_layer) {
+            fprintf(stderr, "%s: [cublas] offloading output layer to GPU\n", __func__);
+        }
+        fprintf(stderr, "%s: [cublas] total VRAM used: %zu MB\n", __func__, vram_total / 1024 / 1024);
+#else
+    (void) n_gpu_layers;
+#endif
+    }
+
     // populate `tensors_by_name`
     for (llama_load_tensor & lt : ml->tensors_map.tensors) {
         model.tensors_by_name.emplace_back(lt.name, lt.ggml_tensor);
     }
 
     ml->load_all_data(progress_callback, progress_callback_user_data, use_mlock ? &lctx.model.mlock_mmap : NULL);
 
-    model.mapping = std::move(ml->mapping);
 #ifdef GGML_USE_CUBLAS
     {
-        const int n_gpu = std::min(n_gpu_layers, int(hparams.n_layer));
+        std::vector<struct ggml_tensor *> tensors;
+        std::vector<size_t> offsets;
+        for (llama_load_tensor & lt : ml->tensors_map.tensors) {
+            if (lt.ggml_tensor->backend != GGML_BACKEND_CUDA) {
+                continue;
+            }
+            tensors.emplace_back(lt.ggml_tensor);
+            LLAMA_ASSERT(lt.shards.size() == 1);
+            offsets.emplace_back(lt.shards.at(0).file_off);
+        }
+        bool cufile_success = ggml_cuda_load_data_cufile(fname.c_str(), tensors.data(), tensors.size(), offsets.data());
 
-        fprintf(stderr, "%s: [cublas] offloading %d layers to GPU\n", __func__, n_gpu);
+        if (!cufile_success) {
+            for (llama_load_tensor & lt : ml->tensors_map.tensors) {
+                if (lt.ggml_tensor->backend != GGML_BACKEND_CUDA) {
+                    continue;
+                }
+                size_t actual_size;
+                void * buf = ggml_cuda_pool_malloc(lt.size, &actual_size);
+                void * buf_host = ggml_cuda_host_malloc(lt.size);
```

Comment:
The `ggml_cuda_pool_malloc` could probably be more efficiently replaced by a `cudaMalloc` directly, no point in keeping these buffers in the pool.

I would also consider keeping this entire branch in `ggml_cuda_load_data_cufile` (maybe renamed to a more generic `ggml_cuda_load_data`).

Also, consider that allocating host memory is sometimes very slow, it may be faster to just use normal memory here.

Reply:
I can't just put this part of the code into `ggml-cuda.cu` if that is what you mean. Reading from disk is done via `llama_file` (which covers a lot of OS-specific problems) and as such cannot be used in ggml code unless we decide that ggml should include both C++ and llama.

Reply:
Actually what I just said may be wrong. In `llama_file` C++ std libraries are used but C seems to have the same functionality for working with files. So I think it would be possible without too much work.